src/rust/cryptography-x509-validation/src/policy/extension.rs

```diff
@@ -171,14 +160,207 @@ impl<B: CryptoOps> ExtensionPolicy<B> {
     }
 }
 
+pub(crate) mod ee {
+    use cryptography_x509::{
+        certificate::Certificate,
+        extensions::{BasicConstraints, ExtendedKeyUsage, Extension, SubjectAlternativeName},
+    };
+
+    use crate::{
+        ops::CryptoOps,
+        policy::{Policy, ValidationError},
+    };
+
+    pub(crate) fn basic_constraints<B: CryptoOps>(
+        _policy: &Policy<'_, B>,
+        _cert: &Certificate<'_>,
+        extn: Option<&Extension<'_>>,
+    ) -> Result<(), ValidationError> {
+        if let Some(extn) = extn {
+            let basic_constraints: BasicConstraints = extn.value()?;
+
+            if basic_constraints.ca {
+                return Err(ValidationError::Other(
+                    "basicConstraints.cA must not be asserted in an EE certificate".to_string(),
+                ));
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn subject_alternative_name<B: CryptoOps>(
+        policy: &Policy<'_, B>,
+        cert: &Certificate<'_>,
+        extn: &Extension<'_>,
+    ) -> Result<(), ValidationError> {
+        match (cert.subject().is_empty(), extn.critical) {
+            // If the subject is empty, the SAN MUST be critical.
+            (true, false) => {
+                return Err(ValidationError::Other(
+                    "EE subjectAltName MUST be critical when subject is empty".to_string(),
+                ));
+            }
+            // If the subject is non-empty, the SAN MUST NOT be critical.
+            (false, true) => {
+                return Err(ValidationError::Other(
+                    "EE subjectAltName MUST NOT be critical when subject is nonempty".to_string(),
+                ))
+            }
+            _ => (),
+        };
+
+        let san: SubjectAlternativeName<'_> = extn.value()?;
+        match policy.subject.matches(&san) {
+            true => Ok(()),
+            false => Err(ValidationError::Other(
+                "EE cert has no matching SAN".to_string(),
+            )),
+        }
+    }
+
+    pub(crate) fn extended_key_usage<B: CryptoOps>(
+        policy: &Policy<'_, B>,
+        _cert: &Certificate<'_>,
+        extn: &Extension<'_>,
+    ) -> Result<(), ValidationError> {
+        let mut ekus: ExtendedKeyUsage<'_> = extn.value()?;
+
+        if ekus.any(|eku| eku == policy.extended_key_usage) {
```

Comment:
How does this handle `2.5.29.37.0` aka `EKU_ANY_KEY_USAGE_OID`?

Reply:
It doesn't at the moment 😅 -- my read of CABF was that `anyExtendedKeyUsage` MUST NOT appear in subscriber certificates. 

Ref:

<img width="552" alt="Screenshot 2023-11-20 at 11 20 59 AM" src="https://github.com/pyca/cryptography/assets/3059210/78287010-5178-4c23-8ffb-306e9a733658">

(Page 80 on CABF v2.0.1)

(This is maybe a place that's worth being not strictly compliant, since my understanding is that lots of PKIs use that EKU?)

Reply:
I'm unsure of how prevalent it is. If it's denied by CABF I'd say just drop a comment about why we do exact match and ignore the any OID for now and move on 😄

Reply:
7b4c2c0ec